Add UI for entities in 3D Viewport

PardhavMaradani opened the pull request:
* Add blender properties corresponding to entities that hold `uuid` and entity type
* Link creation/deletion of session entities to corresponding blender properties
* Add panels for entities listing and trajectory details
* Add tests for all new functionality

PardhavMaradani commented:
I do like the filtering approach because it is less code and maintains data integrity. I think we could do our best for things that we control (like when adding / deleting) entities to ensure the active index is correctly updated. Given the index is of `bpy.data.objects`, which is fairly changeable, we have this unique issue. Maybe this is the small price we pay for data consistency - is is just the numbering and the retaining of the selection. I can try a bit more to see what can be done to minimize these issues.

PardhavMaradani commented:
If I remember correctly, changing props in the `poll` or `draw` context could lead to some exceptions - we ran into some of them in the past.

PardhavMaradani commented:
Just a quick summary of both the approaches:

custom property group: Predictable behavior with respect to selections from the ui list. Entities deleted from Blender (using `x` or from some other API, not ours) lead session entities to get out of sync - this is an existing problem and not something this PR introduces. These session entities get pruned today in `update_entities`.

filtering: Can use `bpy.data.objects` directly and not need separate props for session entities. Has challenges with keeping the selection predictable. Entities deleted from Blender still lead to out of sync in our session object. The ui list will be in sync, but this is something that can be achieved in the above case too.

While writing the above, it seems like we should just stick with the custom property group and the simpler and predictable approach - thought with a bit more property handling code. Please let me know if you agree - the current code has the filtering changes due to the most recent commit, so I can bring them back to the old approach. Thanks

yuxuanzhuang commented:
I got an error when I tried to run this branch with Blender Development addon. It seems to be related to the `register_msgbus_subscriptions`.

Did I do anything wrong here?

<details>

  Exception in module register(): /Users/scottzhuang/Library/Application Support/Blender/4.4/extensions/vscode_development/molecularnodes/__init__.py
  Traceback (most recent call last):
    File "/Applications/Blender.app/Contents/Resources/4.4/scripts/modules/addon_utils.py", line 515, in enable
      mod.register()
    File "/Users/scottzhuang/Library/Application Support/Blender/4.4/extensions/vscode_development/molecularnodes/ui/addon.py", line 69, in register
      register_msgbus_subscriptions()
    File "/Users/scottzhuang/Library/Application Support/Blender/4.4/extensions/vscode_development/molecularnodes/blender/utils.py", line 62, in register_msgbus_subscriptions
      _subscribe_to_msgbus(None)
    File "/Users/scottzhuang/Library/Application Support/Blender/4.4/extensions/vscode_development/molecularnodes/blender/utils.py", line 56, in _subscribe_to_msgbus
      _subscribe_to_active_object_changes()  # subscribe to active object changes
      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/Users/scottzhuang/Library/Application Support/Blender/4.4/extensions/vscode_development/molecularnodes/blender/utils.py", line 46, in _subscribe_to_active_object_changes
      owner=bpy.context.scene.MNSession,
            ^^^^^^^^^^^^^^^^^
  AttributeError: '_RestrictContext' object has no attribute 'scene'
  ADDONS_TO_LOAD [{'load_dir': '/Users/scottzhuang/git_rego/MolecularNodes_GSOC/molecularnodes', 'module_name': 'molecularnodes'}]
  INFO: Removing old link: /Users/scottzhuang/Library/Application Support/Blender/4.4/extensions/vscode_development/molecularnodes
   * Serving Flask app 'Blender Server'
   * Debug mode: on
  Sending: {'type': 'setup', 'blenderPort': 7240, 'debugpyPort': 7713, 'blenderPath': '/Applications/Blender.app/Contents/MacOS/blender', 'scriptsFolder': '/Applications/Blender.app/Contents/MacOS/4.4/scripts', 'addonPathMappings': [{'src': '/Users/scottzhuang/git_rego/MolecularNodes_GSOC/molecularnodes', 'load': '/Users/scottzhuang/Library/Application Support/Blender/4.4/extensions/vscode_development/molecularnodes'}]}
  Waiting for debug client.
  Debug client attached.
  Warning: User Default: [Errno 2] No such file or directory: '/Users/scottzhuang/Library/Application Support/Blender/4.4/extensions/user_default'
  register_class(...): already registered as a subclass 'MN_PT_Scene'
  register_class(...): already registered as a subclass 'MN_UL_EntitiesList'
  register_class(...): already registered as a subclass 'MN_PT_Entities'
  register_class(...): already registered as a subclass 'MN_PT_trajectory'
  register_class(...): already registered as a subclass 'MN_OT_Add_Custom_Node_Group'
  register_class(...): already registered as a subclass 'MN_OT_Residues_Selection_Custom'
  register_class(...): already registered as a subclass 'MN_OT_Assembly_Bio'
  register_class(...): already registered as a subclass 'MN_OT_iswitch_custom'
  register_class(...): already registered as a subclass 'MN_OT_Change_Color'
  register_class(...): already registered as a subclass 'MN_OT_Node_Swap'
  register_class(...): already registered as a subclass 'MN_OT_Import_Fetch'
  register_class(...): already registered as a subclass 'MN_OT_Import_OxDNA_Trajectory'
  register_class(...): already registered as a subclass 'MN_OT_Import_Trajectory'
  register_class(...): already registered as a subclass 'MN_OT_Reload_Trajectory'
  register_class(...): already registered as a subclass 'MN_OT_Import_Map'
  register_class(...): already registered as a subclass 'MN_OT_Import_Star_File'
  register_class(...): already registered as a subclass 'MN_OT_Import_Cell_Pack'
  register_class(...): already registered as a subclass 'MN_OT_Import_Protein_Local'
  register_class(...): already registered as a subclass 'MN_OT_Import_Molecule'
  register_class(...): already registered as a subclass 'MN_FH_Import_Molecule'
  register_class(...): already registered as a subclass 'MolecularNodesObjectProperties'
  register_class(...): already registered as a subclass 'MolecularNodesSceneProperties'
  register_class(...): already registered as a subclass 'TrajectorySelectionItem'
  register_class(...): already registered as a subclass 'MN_UL_TrajectorySelectionListUI'
  register_class(...): already registered as a subclass 'MN_OT_Universe_Selection_Add'
  register_class(...): already registered as a subclass 'MN_OT_Universe_Selection_Delete'
  register_class(...): already registered as a subclass 'MN_OT_Template_Install'
  register_class(...): already registered as a subclass 'MN_OT_Template_Uninstall'
  register_class(...): already registered as a subclass 'MolecularNodesPreferences'
  register_class(...): already registered as a subclass 'MN_OT_Session_Remove_Item'
  register_class(...): already registered as a subclass 'MN_OT_Session_Create_Object'
  register_class(...): already registered as a subclass 'MN_MT_Node'
  register_class(...): already registered as a subclass 'MN_MT_Node_Animate'
  register_class(...): already registered as a subclass 'MN_MT_Node_Assembly'
  register_class(...): already registered as a subclass 'MN_MT_Node_Bonds'
  register_class(...): already registered as a subclass 'MN_MT_Node_Color'
  register_class(...): already registered as a subclass 'MN_MT_Node_Density'
  register_class(...): already registered as a subclass 'MN_MT_Node_DNA'
  register_class(...): already registered as a subclass 'MN_MT_Node_Style'
  register_class(...): already registered as a subclass 'MN_MT_Node_Select'
  register_class(...): already registered as a subclass 'MN_MT_Node_Topology'
  register_class(...): already registered as a subclass 'MN_MT_Node_Attributes'
  register_class(...): already registered as a subclass 'MN_MT_Node_Geometry'
  register_class(...): already registered as a subclass 'MN_MT_Node_Simulation'
  register_class(...): already registered as a subclass 'MN_MT_Node_Curves'
  register_class(...): already registered as a subclass 'MN_MT_Node_Utils'
  register_class(...): already registered as a subclass 'MN_MT_Node_Fields'
  Exception in module register(): /Users/scottzhuang/Library/Application Support/Blender/4.4/extensions/vscode_development/molecularnodes/__init__.py
  Traceback (most recent call last):
    File "/Applications/Blender.app/Contents/Resources/4.4/scripts/modules/addon_utils.py", line 515, in enable
      mod.register()
    File "/Users/scottzhuang/Library/Application Support/Blender/4.4/extensions/vscode_development/molecularnodes/ui/addon.py", line 69, in register
      register_msgbus_subscriptions()
    File "/Users/scottzhuang/Library/Application Support/Blender/4.4/extensions/vscode_development/molecularnodes/blender/utils.py", line 62, in register_msgbus_subscriptions
      _subscribe_to_msgbus(None)
    File "/Users/scottzhuang/Library/Application Support/Blender/4.4/extensions/vscode_development/molecularnodes/blender/utils.py", line 56, in _subscribe_to_msgbus
      _subscribe_to_active_object_changes()  # subscribe to active object changes
      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/Users/scottzhuang/Library/Application Support/Blender/4.4/extensions/vscode_development/molecularnodes/blender/utils.py", line 46, in _subscribe_to_active_object_changes
      owner=bpy.context.scene.MNSession,
            ^^^^^^^^^^^^^^^^^
  AttributeError: '_RestrictContext' object has no attribute 'scene'
  Error: '_RestrictContext' object has no attribute 'scene'
  Traceback (most recent call last):
    File "/Users/scottzhuang/.vscode-insiders/extensions/jacqueslucke.blender-development-0.0.25/pythonFiles/include/blender_vscode/load_addons.py", line 222, in load
      bpy.ops.preferences.addon_enable(module=addon_name)
    File "/Applications/Blender.app/Contents/Resources/4.4/scripts/modules/bpy/ops.py", line 109, in __call__
      ret = _op_call(self.idname_py(), kw)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  RuntimeError: Error: '_RestrictContext' object has no attribute 'scene'
</details>

PardhavMaradani commented:
Hi Yuxuan, I will take a look the error if we decide to go with this approach (filtered `bpy.data.objects`) vs the custom properties. If it is the later, this code won't be involved. I'll make a note to debug  this anyway - seems like there is no `scene` in the context.

PardhavMaradani commented:
Hi Brady, after quite a bit of testing both approaches, I moved back to the old approach (props based) and added further validations, cleanup and filtering to ensure the list is in sync with what shows up in Blender. All of this works fairly well now. Could you please review. Thanks

BradyAJohnston commented:
I've tested it out and it works well. I'm happy to go with your choice here as you've spent the most time with it. An initial look of the code looks good to me. Will have a proper review tomorrow but should be good to go.

yuxuanzhuang commented on molecularnodes/session.py:
```diff
@@ -102,11 +126,22 @@ def prune(self) -> None:
+        pruned = False
+        for i, entity in enumerate(entities):
+            if entity.name not in self.entities:
+                entities.remove(i)  # remove entity from collection
```
You are modifying the list that is being iterated. It's probably gonna lead to unexpected errors.

yuxuanzhuang replied:
is it being tested?

PardhavMaradani replied:
I did test this, but redid this to eliminate any uncertainty in the latest commit

yuxuanzhuang reviewed (changes requested):
I am able to install this version and everything runs smoothly without errors. Really love this UI for entities.

Just two nitpicking
- Would it be possible to select the entity object when clicking on it in the list?
- The filter function in the UI doesn’t seem to be working.

PardhavMaradani commented:
> * Would it be possible to select the entity object when clicking on it in the list?

The latest commit should address this. However, the reverse - ie., updating the entity selected in our ui based on what's selected/active in the outliner will require the msgbus subscription - for the active object changes etc, so that doesn't exist.

> * The filter function in the UI doesn’t seem to be working.

We don't have access to the object name in the entity property collection - just the uuid, to lookup the entity from the session. It is for this reason, I didn't find a way to do the filtering based on name in the `filter_items` for this ui list. I'll make a note of this for the future.

BradyAJohnston commented:
This all looks good for now. Some minor tweaks can be made as we develop this more moving forward.
